csharp/source/autowrap/csharp/csharp.d

```diff
@@ -0,0 +1,1061 @@
+        } else {
+            funcStr ~= "        public static %s %s(".format(getCSharpInterfaceType(returnTypeStr), methodName);
+        }
+        static foreach(pc; 0..paramNames.length) {
+            if (is(paramTypes[pc] == string) || is(paramTypes[pc] == wstring) || is(paramTypes[pc] == dstring)) {
+                funcStr ~= getCSharpInterfaceType(fullyQualifiedName!(paramTypes[pc])) ~ " " ~ paramNames[pc] ~ ", ";
+            } else if (isArray!(paramTypes[pc])) {
+                funcStr ~= "Range<" ~ getCSharpInterfaceType(fullyQualifiedName!(paramTypes[pc])) ~ "> " ~ paramNames[pc] ~ ", ";
+            } else {
+                funcStr ~= getCSharpInterfaceType(fullyQualifiedName!(paramTypes[pc])) ~ " " ~ paramNames[pc] ~ ", ";
+            }
+        }
+        if (paramNames.length > 0) {
+            funcStr = funcStr[0..$-2];
+        }
+        funcStr ~= ") {" ~ newline;
+        if (isNothrow && is(returnType == void)) {
+            funcStr ~= "            dlang_%s(".format(methodName);
+        } else { 
+            funcStr ~= "            var dlang_ret = dlang_%s(".format(methodName);
+        }
+        static foreach(pc; 0..paramNames.length) {
+            if (is(paramTypes[pc] == string)) {
+                funcStr ~= "SharedFunctions.CreateString(" ~ paramNames[pc] ~ "), ";
+            } else if (is(paramTypes[pc] == wstring)) {
+                funcStr ~= "SharedFunctions.CreateWString(" ~ paramNames[pc] ~ "), ";
+            } else if (is(paramTypes[pc] == dstring)) {
+                funcStr ~= "SharedFunctions.CreateDString(" ~ paramNames[pc] ~ "), ";
+            } else if (isArray!(paramTypes[pc])) {
+                funcStr ~= paramNames[pc] ~ ".ToSlice(), ";
+            } else {
+                funcStr ~= paramNames[pc] ~ ", ";
+            }
+        }
+        if (paramNames.length > 0) {
+            funcStr = funcStr[0..$-2];
+        }
+        funcStr ~= ");" ~ newline;
+        if (is(returnType == void)) {
+            funcStr ~= "            dlang_ret.EnsureValid();" ~ newline;
+        } else {
+            if (is(returnType == string)) {
+                funcStr ~= "            return SharedFunctions.SliceToString(dlang_ret, DStringType._string);" ~ newline;
+            } else if (is(returnType == wstring)) {
+                funcStr ~= "            return SharedFunctions.SliceToString(dlang_ret, DStringType._wstring);" ~ newline;
+            } else if (is(returnType == dstring)) {
+                funcStr ~= "            return SharedFunctions.SliceToString(dlang_ret, DStringType._dstring);" ~ newline;
+            } else if (isArray!(returnType)) {
+                funcStr ~= "            return new Range<%s>(dlang_ret);".format(getCSharpInterfaceType(returnTypeStr)) ~ newline;
+            } else {
+                funcStr ~= "            return dlang_ret;" ~ newline;
+            }
+        }
+        funcStr ~= "        }" ~ newline;
+        ns.functions ~= funcStr;
+    }
```

Comment:
maybe I'm being too fussy, but this would be way easier to understand if it was structured semantically not linearly, i.e. the different bits of code-gen for different parts of the generated function (attributes, signature, core) were done in separate functions and then pieced together here.

Or perhaps even just some comments breaking it in to sections would help